pandas/tests/copy_view/test_setitem.py

```diff
@@ -0,0 +1,95 @@
+import numpy as np
+
+from pandas import (
+    DataFrame,
+    Index,
+    RangeIndex,
+    Series,
+)
+import pandas._testing as tm
+
+# -----------------------------------------------------------------------------
+# Copy/view behaviour for the values that are set in a DataFrame
+
+
+def test_set_column_with_array():
+    # Case: setting an array as a new column (df[col] = arr) copies that data
+    df = DataFrame({"a": [1, 2, 3], "b": [4, 5, 6]})
+    arr = np.array([1, 2, 3], dtype="int64")
+
+    df["c"] = arr
+
+    # the array data is copied
+    assert not np.shares_memory(df["c"].values, arr)
+    # and thus modifying the array does not modify the DataFrame
+    arr[0] = 0
+    tm.assert_series_equal(df["c"], Series([1, 2, 3], name="c"))
+
+
+def test_set_column_with_series(using_copy_on_write):
+    # Case: setting a series as a new column (df[col] = s) copies that data
+    df = DataFrame({"a": [1, 2, 3], "b": [4, 5, 6]})
+    s = Series([1, 2, 3])
+
+    df["c"] = s
+
+    if using_copy_on_write:
+        # with CoW we can delay the copy
+        assert np.shares_memory(df["c"].values, s.values)
+    else:
+        # the series data is copied
+        assert not np.shares_memory(df["c"].values, s.values)
+
+    # and modifying the series does not modify the DataFrame
+    s.iloc[0] = 0
```

Comment:
could also assert `s.iloc[0] == 0`, as failure for that to hold would be another way the next assertion could pass